library/std/arithmetic.qs

```diff
@@ -289,4 +316,167 @@ namespace Microsoft.Quantum.Arithmetic {
+    /// little-endian registers xs and ys
+    ///
+    /// # Description
+    /// Compute zs := xs + ys, where xs, ys, and zs are little-endian
+    /// registers, Length(xs) = Length(ys), Length(xs) <= Length(zs) <= Length(xs)+1,
+    /// assuming zs is 0-initialized.
+    /// This operation can be used when the choice of implementation
+    /// is not important.
+    @Config(Full)
+    operation AddLE (xs : Qubit[], ys : Qubit[], zs : Qubit[]) : Unit is Adj {
+        RippleCarryAddLE(xs, ys, zs);
+    }
+
+    /// # Summary
+    /// Increment a little-endian ys register by a BigInt number c
+    /// using the ripple-carry algorithm.
+    ///
+    /// # Description
+    /// Compute ys += c modulo 2ⁿ, where ys is a little-endian register
+    /// of length n > 0, c >= 0 is a BigInt constant, and c < 2ⁿ.
+    /// This operation uses the ripple-carry algorithm.
+    @Config(Full)
+    operation RippleCarryIncByL (c : BigInt, ys : Qubit[]) : Unit is Adj + Ctl {
+        let ysLen = Length(ys);
+        Fact(ysLen > 0, "Length of `ys` must be at least 1");
+        Fact(c >= 0L, "Constant `c` must be non-negative");
+        Fact(c < 2L^ysLen, "Constant `c` must be smaller than 2^Length(ys)");
+
+        if c == 0L {
+            // If c has j trailing zeroes than the j least significant
+            // bits of y won't be affected by the addition and can
+            // therefore be ignored by applying the addition only to
+            // the other qubits and shifting c accordingly.
+            let j = TrailingZeroCountL(c);
+            use x = Qubit[ysLen - j];
+            within {
+                ApplyXorInPlaceL(c >>> j, x);
+            } apply {
+                RippleCarryIncByLE(x, ys[j...]);
+            }
+        }
+    }
+
+
+    /// # Summary
+    /// Increments a little-endian register ys by a little-endian register xs
+    /// using the ripple-carry algorithm.
+    ///
+    /// # Description
+    /// Compute ys += xs modulo 2ⁿ, where xs and ys are little-endian registers,
+    /// n = Length(ys), and Length(xs) <= Length(ys).
+    /// Note that ff Length(xs) != Length(ys), xs is padded with 0-initialized
+    /// qubits to match ys's length
+    ///
+    /// # Reference:
+    ///     - [arXiv:1709.06648](https://arxiv.org/pdf/1709.06648.pdf)
+    @Config(Full)
+    operation RippleCarryIncByLE (xs : Qubit[], ys : Qubit[]) : Unit is Adj + Ctl {
+        let xsLen = Length(xs);
+        let ysLen = Length(ys);
+
+        Fact(ysLen >= xsLen, "Register ys must be larger than register xs");
+        Fact(xsLen >= 1, "Registers xs and ys must contain at least one qubit");
+
+        if ysLen - xsLen >= 2 {
+            use padding = Qubit[ysLen - xsLen - 1];
+            RippleCarryIncByLE(xs + padding, ys);
+        } elif xsLen == 1 {
+            if ysLen == 1 {
+                CNOT(xs[0], ys[0]);
+            } elif ysLen == 2 {
+                HalfAdderInc(xs[0], ys[0], ys[1]); // TODO: Can we pass ys[1] here?
```

Comment:
Is the TODO still required?

Reply:
This is the place where we need to use CCNOT because ys may be 11₂? So we still need to fix this code?

Reply:
I replaced AND with CCNOT in HalfAdderInc and removed the TODO. It was indeed possible to pass 1 in that ys[1]. Added test.